packages/gatsby/src/datastore/lmdb/query/suggest-index.ts

```diff
@@ -0,0 +1,150 @@
+  dbQueryToDottedField,
+  getFilterStatement,
+  prepareQueryArgs,
+  sortBySpecificity,
+} from "../../common/query"
+import { isDesc } from "./common"
+
+interface ISelectIndexArgs {
+  filter: IRunQueryArgs["queryArgs"]["filter"]
+  sort: IRunQueryArgs["queryArgs"]["sort"]
+  maxFields?: number
+}
+
+type IndexField = [fieldName: string, orderDirection: number]
+type IndexFields = Array<IndexField>
+
+export function suggestIndex({
+  filter,
+  sort,
+  maxFields = 4,
+}: ISelectIndexArgs): Array<IndexField> {
+  const filterQueries = createDbQueriesFromObject(prepareQueryArgs(filter))
+  const filterQueriesThatCanUseIndex = getQueriesThatCanUseIndex(filterQueries)
+  const sortFields: Array<IndexField> = getSortFieldsThatCanUseIndex(sort)
+
+  if (!sortFields.length && !filterQueriesThatCanUseIndex.length) {
+    return []
+  }
+  if (!filterQueriesThatCanUseIndex.length) {
+    return dedupeAndTrim(sortFields, maxFields)
+  }
+  if (!sortFields.length) {
+    return dedupeAndTrim(toIndexFields(filterQueriesThatCanUseIndex), maxFields)
+  }
+  const overlap = findOverlappingFields(
+    filterQueriesThatCanUseIndex,
+    sortFields
+  )
+  if (!overlap.size) {
+    // Filter and sort fields do not overlap.
+    // In this case combined index for filter+sort only makes sense when all filters have `eq` predicate
+    // Same as https://docs.mongodb.com/manual/tutorial/sort-results-with-indexes/#sort-and-non-prefix-subset-of-an-index
+    const eqFields = getFieldsWithEqPredicate(filterQueriesThatCanUseIndex)
+
+    if (eqFields.size === filterQueriesThatCanUseIndex.length) {
+      const eqFilterFields: Array<IndexField> = [...eqFields].map(f => [f, 1])
+      return dedupeAndTrim([...eqFilterFields, ...sortFields], maxFields)
+    }
+    // Single unbound range filter: "lt", "gt", "gte", "lte" (but not "in"): prefer sort fields
+    if (
+      filterQueriesThatCanUseIndex.length === 1 &&
+      getFilterStatement(filterQueriesThatCanUseIndex[0]).comparator !==
+        DbComparator.IN
+    ) {
+      return dedupeAndTrim(sortFields, maxFields)
+    }
+    // Otherwise prefer filter fields
+    return dedupeAndTrim(toIndexFields(filterQueriesThatCanUseIndex), maxFields)
+  }
+
+  // There is an overlap:
+  // First add all non-overlapping filter fields to index prefix, then all sort fields (including overlapping)
+  const filterFields = filterQueriesThatCanUseIndex
+    .map((q): IndexField => [dbQueryToDottedField(q), 1])
+    .filter(([name]) => !overlap.has(name))
+
+  return dedupeAndTrim([...filterFields, ...sortFields], maxFields)
+}
+
+/**
+ * Returns queries that can potentially use index.
+ * Returned list is sorted by query specificity
+ */
+function getQueriesThatCanUseIndex(all: Array<DbQuery>): Array<DbQuery> {
+  const canUseIndex = new Set([
+    DbComparator.EQ,
+    DbComparator.IN,
+    DbComparator.GTE,
+    DbComparator.LTE,
+    DbComparator.GT,
+    DbComparator.LT,
+    DbComparator.NIN,
+    DbComparator.NE,
+  ])
```

Comment:
NIT: this set doesn't have to be reconstructed every time. Also seems like exact same array is declared separately in packages/gatsby/src/datastore/common/query.ts

Reply:
It was a single set initially. But then I've realized that `NIN` and `NE` could be used conditionally - single key indexes can use those, multi-key indexes cannot. 

They are identical at the moment simply because `suggestIndex` doesn't "know" if the index will be multiKey or singleKey. Otherwise, this set would be different depending on "type" of the index.

But the specificity never changes. So those two sets are essentially different things. They have different reasons to change and that's why I've split them intentionally.

But moving the set outside of this function makes sense. Will do 👍

Reply:
Addressed in https://github.com/gatsbyjs/gatsby/pull/32261/commits/b6a8c581b768cc49aa843216cccba1eee999859f

Reply:
Addressed in https://github.com/gatsbyjs/gatsby/pull/32261/commits/b6a8c581b768cc49aa843216cccba1eee999859f